ml_perf/reference_implementation.py

```diff
@@ -0,0 +1,279 @@
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Runs a reinforcement learning loop to train a Go playing model."""
+
+import sys
+
+sys.path.insert(0, '.')
+
+import logging
+import numpy
+import os
+import random
+import re
+import shutil
+import subprocess
+import tensorflow
+import utils
+
+from absl import app, flags
+from rl_loop import example_buffer, fsdb, shipname
+
+flags.DEFINE_string('engine', 'tf', 'Engine to use for inference.')
+
+FLAGS = flags.FLAGS
+
+
+class State:
+
+  _NAMES = ['bootstrap'] + random.Random(0).sample(shipname.NAMES,
+                                                   len(shipname.NAMES))
+
+  def __init__(self):
+    self.iter_num = 0
+    self.play_model_num = 0
+    self.play_model_name = self.play_output_name
+    self.train_model_num = 1
+
+  @property
+  def play_output_name(self):
+    return '%06d-%s' % (self.iter_num, self._NAMES[self.play_model_num])
+
+  @property
+  def play_model_path(self):
+    return os.path.join(fsdb.models_dir(), self.play_model_name)
+
+  @property
+  def train_model_name(self):
+    return '%06d-%s' % (self.iter_num, self._NAMES[self.train_model_num])
+
+  @property
+  def train_model_path(self):
+    return os.path.join(fsdb.models_dir(), self.train_model_name)
+
+  @property
+  def seed(self):
+    return self.iter_num + 1
+
+
+def checked_run(cmd, name):
+  logging.info('Running %s:\n  %s', name, '\n  '.join(cmd))
+  with utils.logged_timer('%s finished' % name.capitalize()):
+    completed_process = subprocess.run(
+        cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+    if completed_process.returncode:
+      logging.error('Error running %s: %s', name,
+                    completed_process.stdout.decode())
+      raise RuntimeError('Non-zero return code executing %s' % ' '.join(cmd))
+  return completed_process
+
+
+def get_lines(completed_process, slice):
+  return '\n'.join(completed_process.stdout.decode()[:-1].split('\n')[slice])
+
+
+class MakeSlice(object):
+
+  def __getitem__(self, item):
+    return item
+
+
+make_slice = MakeSlice()
+
+
+def cc_flags(state):
+  return [
+      '--engine={}'.format(FLAGS.engine),
+      '--virtual_losses=8',
```

Comment:
I think this should be a constant with explanation at the top as it's one factor that will vary with GPU/CPU/TPU

Reply:
added a TODO